indykite/resource_consent.go

```diff
@@ -0,0 +1,132 @@
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package indykite
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/structure"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/indykite/indykite-sdk-go/config"
+	configpb "github.com/indykite/indykite-sdk-go/gen/indykite/config/v1beta1"
+)
+
+const (
+	consentPurposeKey        = "purpose"
+	consentApplicationIDKey  = "application_id"
+	consentValidityPeriodKey = "validity_period"
+	consentRevokeAfterUseKey = "revoke_after_use"
+	consentDataPointsKey     = "data_points"
+)
+
+func resourceConsent() *schema.Resource {
+	readContext := configReadContextFunc(resourceConsentFlatten)
+
+	return &schema.Resource{
+		CreateContext: configCreateContextFunc(resourceConsentBuild, readContext),
+		ReadContext:   readContext,
+		UpdateContext: configUpdateContextFunc(resourceConsentBuild, readContext),
+		DeleteContext: configDeleteContextFunc(),
+		Importer: &schema.ResourceImporter{
+			StateContext: basicStateImporter,
+		},
+
+		Timeouts: defaultTimeouts(),
+		Schema: map[string]*schema.Schema{
+			locationKey:    locationSchema(),
+			customerIDKey:  setComputed(customerIDSchema()),
+			appSpaceIDKey:  setComputed(appSpaceIDSchema()),
+			nameKey:        nameSchema(),
+			displayNameKey: displayNameSchema(),
+			descriptionKey: descriptionSchema(),
+			createTimeKey:  createTimeSchema(),
+			updateTimeKey:  updateTimeSchema(),
+
+			consentPurposeKey: {
+				Type:         schema.TypeString,
+				Required:     true,
+				Description:  "Purpose of the consent",
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			consentApplicationIDKey: {
+				Type:             schema.TypeString,
+				Required:         true,
+				Description:      "Application ID",
+				ValidateDiagFunc: ValidateGID,
+			},
+			consentValidityPeriodKey: {
+				Type:        schema.TypeInt,
+				Required:    true,
+				Description: "Specifies the duration in second that the consent remains valid, ranging from 1 day to 2 years",
+				ValidateFunc: func(val any, key string) ([]string, []error) {
+					var errs []error
+					v := val.(int)
+					if v < 86400 || v > 63072000 {
+						errs = append(errs,
+							fmt.Errorf("%q must be between 86400 (1 day) and 63072000 (2 years) seconds, got: %d",
+								key,
+								v))
+					}
+					return nil, errs
+				},
```

Comment:
Why you have custom validation function? Why not just use `validation.IntBetween(86400, 63072000)`?